components/Ripple/src/MDCStatefulRippleView.m

```diff
@@ -0,0 +1,246 @@
+    // we want to return early so we don't add multiple selected overlays, as there can be only one.
+    return;
+  }
+  _selected = selected;
+  // Go into the selected state visually.
+  if (selected) {
+    if (!self.activeRippleLayer) {
+      // If we go into the selected state but a ripple layer doesn't exist yet, it means we went
+      // into this state without initially creating the ripple overlay by going through the
+      // highlighted state. This usually occurs when cells are reused and the selected state is
+      // manually set to show the cell's existing state.
+      [self updateRippleColor];
+      [self beginRippleTouchDownAtPoint:_lastTouch animated:NO completion:nil];
+    } else {
+      [self updateActiveRippleColor];
+    }
+  } else {
+    // If we are no longer selecting, we cancel all the ripples.
+    [self updateRippleColor];
+    [self cancelAllRipplesAnimated:YES completion:nil];
+  }
+}
+
+- (void)setRippleHighlighted:(BOOL)rippleHighlighted {
+  if (rippleHighlighted == _rippleHighlighted) {
+    return;
+  }
+  _rippleHighlighted = rippleHighlighted;
+  // Go into the highlighted state visually.
+  if (rippleHighlighted && !_tapWentInsideOfBounds) {
+    // If ripple becomes highlighted we initiate a ripple with animation.
+    [self updateRippleColor];
+    [self beginRippleTouchDownAtPoint:_lastTouch animated:_didReceiveTouch completion:nil];
+  } else {
+    if (!self.allowsSelection && !self.dragged && !_tapWentOutsideOfBounds) {
+      // We dissolve the ripple when highlighted is NO, unless we are going into
+      // selection or dragging.
+      [self updateRippleColor];
+      [self beginRippleTouchUpAnimated:YES completion:nil];
+    }
+  }
+}
+
+- (void)setDragged:(BOOL)dragged {
+  if (dragged == _dragged) {
+    return;
+  }
+  _dragged = dragged;
+  // Go into the dragged state visually.
+  if (dragged) {
+    if (!self.activeRippleLayer) {
+      // If we go into the dragged state manually, without coming from the highlighted state,
+      // We present the ripple overlay instantly without animation.
+      [self updateRippleColor];
+      [self beginRippleTouchDownAtPoint:_lastTouch animated:NO completion:nil];
+    } else {
+      [self updateActiveRippleColor];
+    }
+  } else {
+    // If we are no longer dragging, we cancel all the ripples.
+    [self updateRippleColor];
+    [self cancelAllRipplesAnimated:YES completion:nil];
+  }
+}
+
+- (BOOL)pointInsideSuperview:(CGPoint)point withEvent:(UIEvent *)event {
+  CGPoint superviewPoint = [self convertPoint:point toView:self.superview];
+  return [self.superview pointInside:superviewPoint withEvent:event];
+}
+
+- (BOOL)pointInside:(CGPoint)point withEvent:(UIEvent *)event {
+  if ([self pointInsideSuperview:point withEvent:event]) {
+    _didReceiveTouch = YES;
+    _lastTouch = point;
+    _tapWentInsideOfBounds = NO;
+    _tapWentOutsideOfBounds = NO;
+  }
```

Comment:
Please document why we need to implement this side effect behavior.

Reply:
Done.